### PR TITLE
PYTHON-3716 OIDC-SASL Follow-Up

### PR DESCRIPTION
**Pull request by blink1073**
- [x] Use updated secrets handling from https://github.com/mongodb-labs/drivers-evergreen-tools/pull/345
- [x] Use dedicated Atlas hosts for testing
- [x] Move global cache to the mongo client level and update tests
- [x] Clean up reauth logic in preparation for DRIVERS-2672
- [x] Test on macos and Windows - requires https://github.com/mongodb-labs/drivers-evergreen-tools/pull/349
- [x] Enforce timeout on callback - we decided against this.
- [x] Update specifications

### CODE REVIEW

**Comment by blink1073** [comment]
Working locally with the Atlas hosts:

```
================================= 32 passed, 1 skipped, 1 warning in 71.53s (0:01:11) ==================================
```

**Comment by blink1073** [comment]
cc @durran

File: test/auth_aws/test_auth_oidc.py [hunk 1]
```diff
@@ -1,828 +0,0 @@
-# Copyright 2023-present MongoDB, Inc.
```
Comment (ShaneHarvey) [inline test/auth_aws/test_auth_oidc.py:1]: Would you be able to rename this file in a separate commit to make this review easier?

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: .evergreen/run-tests.sh [hunk 2]
```diff
@@ -224,7 +227,18 @@ fi
 
 if [ -n "$TEST_AUTH_OIDC" ]; then
     python -m pip install ".[aws]"
-    TEST_ARGS="test/auth_aws/test_auth_oidc.py"
+
+    # Work around for root certifi not being installed.
+    # TODO: Remove after PYTHON-3952
```
[the remark this answers is not in the document]
Comment (blink1073) [reply .evergreen/run-tests.sh]: Updated to say "is deployed", I still need to coordinate that.

File: test/auth_aws/test_auth_oidc.py (hunk 1 above)
Comment (blink1073) [reply test/auth_aws/test_auth_oidc.py]: Done in #1378

**Review by NoahStapp — CHANGES_REQUESTED** [review]
A couple minor comments, otherwise looks great!

File: pymongo/auth_oidc.py [hunk 3]
```diff
@@ -247,37 +173,59 @@ def auth_start_cmd(self, use_callbacks: bool = True) -> Optional[SON[str, Any]]:
...
-            # Check if we've already changed tokens.
-            if prev_id == self.token_gen_id:
-                self.reauth_gen_id = self.idp_info_gen_id
-                self.token_exp_utc = None
-                if not self.properties.refresh_token_callback:
-                    self.clear()
+    def reauthenticate(self, conn: Connection) -> Optional[Mapping[str, Any]]:
+        """Handle a reauthenticate from the server."""
+        # First see if we have the a newer token on the authenticator.
+        prev_id = getattr(conn, "oidc_token_gen_id", 0)
```
Comment (ShaneHarvey) [inline pymongo/auth_oidc.py:188]: Let's make oidc_token_gen_id a well defined property on Connection and avoid getattr.

File: .evergreen/run-mongodb-oidc-test.sh [hunk 4]
```diff
@@ -1,56 +1,48 @@
...
-export MONGODB_URI="$MONGODB_URI"
-
-echo $MONGODB_URI_SINGLE
-echo $MONGODB_URI_MULTIPLE
-echo $MONGODB_URI
+# Get the drivers secrets.  Use an existing secrets file first.
+if [ ! -f "./secrets-export.sh" ]; then
+    bash .evergreen/tox.sh -m aws-secrets -- drivers/oidc
+fi
+source ./secrets-export.sh
```
[the remark this answers is not in the document]
Comment (blink1073) [reply .evergreen/run-mongodb-oidc-test.sh:18]: Done

File: .evergreen/run-tests.sh [hunk 5]
```diff
@@ -255,6 +269,9 @@ fi
 # Show the installed packages
 PIP_QUIET=0 python -m pip list
 
+python -c "import urllib.request;urllib.request.urlopen('https://www.google.com')"
```
[the remark this answers is not in the document]
Comment (blink1073) [reply .evergreen/run-tests.sh:272]: That was for debugging, fixed

File: pymongo/auth_oidc.py [hunk 6]
```diff
@@ -92,150 +67,101 @@ def _get_authenticator(
...
-            now_utc = datetime.now(timezone.utc)
-            exp_utc = self.token_exp_utc
-            buffer_seconds = TOKEN_BUFFER_MINUTES * 60
-            if (exp_utc - now_utc).total_seconds() >= buffer_seconds:
-                current_valid_token = True
-
-        timeout = CALLBACK_TIMEOUT_SECONDS
-        if not use_callbacks and not current_valid_token:
+        # TODO: DRIVERS-2672, handle machine callback here as well.
+        cb = properties.request_token_callback
```
[the remark this answers is not in the document]
Comment (blink1073) [reply pymongo/auth_oidc.py:89]: Done

File: pymongo/auth_oidc.py (hunk 3 above)
Comment (blink1073) [reply pymongo/auth_oidc.py:188]: Ah, it was already there as an optional int, updated

**Review by ShaneHarvey — APPROVED** [review]

**Review by NoahStapp — APPROVED** [review]

**Pull request merged**